featureflags/evaluations/evaluator.py

```diff
@@ -0,0 +1,303 @@
+        log.debug("Sorted serving rules %s", rules)
+
+        identifier = None
+        for rule in rules:
+            # if evaluation is false just continue to next rule
+            if not self._evaluate_rule(rule, target):
+                log.debug(
+                    "Unsuccessful evaluation of rule %s continue to next rule",
+                    rule)
+                continue
+
+            # rule matched, check if there is distribution
+            distribution = rule.serve.distribution
+            if not isinstance(distribution, Unset) and distribution:
+                log.debug("Evaluate distribution %s", distribution)
+                identifier = self._evaluate_distribution(
+                    distribution, target)
+
+            # rule matched, here must be variation if distribution is None
+            variation = rule.serve.variation
+            if not isinstance(variation, Unset) and variation:
+                log.debug("Return rule variation identifier %s", identifier)
+                identifier = variation
+
+            return identifier
+
+        log.debug("All rules failed, return empty variation identifier")
+        return None
+
+    def _evaluate_variation_map(self, var_target_map: List[VariationMap],
+                                target: Target) -> Optional[str]:
+        if not target or not var_target_map:
+            log.debug("Target is none")
+            return None
+
+        for variation_map in var_target_map:
+            if not isinstance(variation_map.targets, Unset) and next(
+                (val for val in variation_map.targets
+                    if not isinstance(val, Unset) and val.identifier ==
+                 target.identifier), None) is not None:
+                log.debug("Evaluate variation map with result %s",
+                          variation_map.variation)
+                return variation_map.variation
+
+            segment_identifiers = variation_map.target_segments
+            if not isinstance(segment_identifiers, Unset) and \
+                    self._check_target_in_segment(segment_identifiers, target):
+                log.debug("Evaluate variationMap with segment " +
+                          "identifiers % s and return % s",
+                          segment_identifiers, variation_map.variation)
+                return variation_map.variation
+
+        return None
+
+    def _evaluate_flag(self, fc: FeatureConfig,
+                       target: Target) -> Optional[Variation]:
+        variation: Optional[str] = fc.off_variation
+        log.debug("feature %s state is %s", fc.feature, fc.state)
+        if fc.state == FeatureState.ON:
+            variation = None
+            if not isinstance(fc.variation_to_target_map, Unset):
+                variation = self._evaluate_variation_map(
+                    fc.variation_to_target_map, target)
+                log.debug("variation %s found in target map", variation)
+
+            if not variation:
```

Comment:
I think this would be cleaner and more concise as:
```
if not variation:
    if not isinstance(fc.default_serve.distribution,
                      Unset):
        variation = self._evaluate_distribution(
            fc.default_serve.distribution, target)
        log.debug(
            "variation %s found in default serve distribution",
             variation)
    elif not isinstance(fc.default_serve.variation,
                        Unset):
        variation = fc.default_serve.variation
        log.debug("variation %s found in default serve", variation)
    else
        variation = self._evaluate_rules(fc.rules, target)
        log.debug("variation %s found in rules", variation)
```

Reply:
I did this on one liner with `or` but it was not readable. I choose to go on every step of evaluation and I need to check on every method call if variation have a value, I am not sure if this logic is the same as our steps of evaluating flag:
1. Check Evaluate variation map and return variation
2. if returned variation is None then Evaluate rule
3. if returned variation is None then Evaluate distribution
4. if returned variation is None then Return default serve variation
5. findVariation if not variation serve offVariation